### PR TITLE
[IMP] account, purchase_requisition: product catalog

### PR DESCRIPTION
**Pull request by malb-odoo**
A new catalog feature was added on Sales, Purchase and Field services. To be compliant with the other modules, it needed to be imported to account.

In product a mixins was created to be able to extend the code to other modules. This commit will override the mixins method to add the catalog feature to account move.

When being on a bill (with a partner set), a filter is put by default.

task: 3635565

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase_requisition/models/product.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ class ProductProduct(models.Model):
 
     def _prepare_sellers(self, params=False):
         sellers = super(ProductProduct, self)._prepare_sellers(params=params)
-        if params and params.get('order_id'):
+        if params and params.get('order_id') and params['order_id']._fields.get("requisition_id"):
```
Comment (ticodoo) [inline addons/purchase_requisition/models/product.py:19]: Why this change? If an order is passed in the param and the order doesn't have a requisition_id then we want to make sure that we don't select a seller that DOES have purchase_requisition_id, otherwise we may incorrectly select a seller that we shouldn't

File: addons/product/static/src/product_catalog/kanban_controller.js [hunk 2]
```diff
@@ -25,11 +25,17 @@ export class ProductCatalogKanbanController extends KanbanController {
...
         const orderIsQuotation = ["draft", "sent"].includes(orderStateInfo[0].state);
-        if (orderIsQuotation) {
+        if (orderStateInfo[0]?.move_type ==="out_invoice") {
+            this.buttonString = _t("Back to Invoice");
+        } else if (orderStateInfo[0]?.move_type ==="in_invoice") {
+            this.buttonString = _t("Back to Bill");
+        } else if (orderIsQuotation) {
             this.buttonString = _t("Back to Quotation");
         } else {
             this.buttonString = _t("Back to Order");
```
Comment (ticodoo) [inline addons/product/static/src/product_catalog/kanban_controller.js:41]: It would be cleaner to make the model specific checks be in overrided functions extended in their specific module. We left it this way during the because Sales Orders and Purchase Orders use the same terms and worked for both when the mixin was moved to the product module. You should never refer to models that aren't included in the module (i.e. `account.move`). Until other modules reuse this mixin, I think it's fine to leave this as the default behavior in product and override it only in account

**Review by ticodoo — COMMENTED** [review]
I didn't look at any of the account changes

File: addons/purchase_requisition/models/product.py (hunk 1 above)
Comment (malb-odoo) [reply addons/purchase_requisition/models/product.py:19]: Since i'm passing an account.move as the order, account.move doesn't have a requisition_id and it causes the thing to crash, i don't think we want to add the requisition_id to our model that's why i've done this change. If you think there is another way i would gladly take it :smile:

File: addons/product/static/src/product_catalog/kanban_controller.js (hunk 2 above)
Comment (malb-odoo) [reply addons/product/static/src/product_catalog/kanban_controller.js:41]: Yeah i thought of making an override for the function but since it was used for both purchase and sale, i decided to it this way but you are right that reference a model outside the model itself is not the best idea, i will change that after my holiday :smile:

File: addons/purchase_requisition/models/product.py (hunk 1 above)
Comment (ticodoo) [reply addons/purchase_requisition/models/product.py:19]: Ah sorry, I misread what you were doing :see_no_evil: please disregard my comment

**Review by chevalierv — APPROVED** [review]
Hello @malb-odoo 

Thanks for your work on this. LGTM :+1: 

Have a nice day :tada:

File: addons/purchase_requisition/models/product.py (hunk 1 above)
Comment (svs-odoo) [reply addons/purchase_requisition/models/product.py:19]: @malb-odoo But does it make sense at the first place to call `_select_seller` while passing an `account.move` as the order ?

**Comment by svs-odoo** [comment]
About your commit message, note than `[ADD]` is used [when adding a new module](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#tag-and-module-name), you should label it `[IMP]` instead

File: addons/purchase_requisition/models/product.py (hunk 1 above)
Comment (malb-odoo) [reply addons/purchase_requisition/models/product.py:19]: @svs-odoo Actually yes it does, in case of a bill, we can have multiple price depending on the seller

File: addons/account/models/account_move.py
Comment (FlorianGilbert) [inline addons/account/models/account_move.py:1]: I've the feeling that a lot of codes could be moves into the `account.move.line` model, why do we decide to do it in `account.move`?
I probably miss something

File: addons/account/models/account_move_line.py [hunk 3]
```diff
@@ -3232,6 +3232,59 @@ def action_automatic_entry(self, default_action=None):
...
 
+    def action_add_from_catalog(self):
+        """ Will open the catalog view """
+        move = self.env['account.move'].browse(self.env.context.get('order_id'))
+        return move.action_add_from_catalog()
+
+    # -------------------------------------------------------------------------
+    # Catalog
+    # -------------------------------------------------------------------------
+    def _get_product_catalog_lines_data(self, **kwargs):
```
Comment (FlorianGilbert) [inline addons/account/models/account_move_line.py:3243]: For me, the code could be simplified. 
We can probably extends the behavior of `_is_readonly` in the right model to avoid arbitrary value in the elif.
The price and the quantity could be computed in the same way by refactoring a bit the code.

**Review by FlorianGilbert — CHANGES_REQUESTED** [review]
Technical review, probably some stupid questions!

File: addons/account/models/account_move.py [hunk 4]
```diff
@@ -2048,6 +2049,84 @@ def _validate_taxes_country(self):
...
+
+    def _update_order_line_info(self, product_id, quantity, **kwargs):
+        """ Update account_move_line information for a given product or create a
+        new one if none exists yet.
+        :param int product_id: The product, as a `product.product` id.
+        :return: The unit price of the product, based on the pricelist of the
+                 sale order and the quantity selected.
+        :rtype: float
+        """
+        move_line = self.line_ids.filtered(lambda line: line.product_id.id == product_id)
```
[the remark this answers is not in the document]
Comment (malb-odoo) [reply addons/account/models/account_move.py:2112]: Nice catch ! There was a traceback caused by the return when the self is bigger then 1 element :smile:

File: addons/account/models/account_move.py [hunk 5]
```diff
@@ -2048,6 +2049,84 @@ def _validate_taxes_country(self):
...
+        if move_line:
+            if quantity != 0:
+                move_line.quantity = quantity
+            elif self.state in ['draft', 'sent']:
+                price_unit = self._get_product_price_and_data(move_line.product_id)['price']
+                move_line.unlink()
```
[the remark this answers is not in the document]
Comment (malb-odoo) [reply addons/account/models/account_move.py:2118]: @chevalierv Any reason the unlink is necessary, i've just copied from purchase and sale module but i just tested it without and it seems to work fine

[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/account/models/account_move.py]: The catalog is designed to allow the user to select products quickly. Therefore, sometimes they may select the wrong product or decide to remove some of them from the quotation. As a customer, you don't want lines of products with a quantity of 0 in a draft sale order.

I would advise keeping the same behavior unless you have a really good reason to keep the lines in your draft invoices.

[the remark this answers is not in the document]
Comment (malb-odoo) [reply addons/account/models/account_move.py]: Fine by me i will just write a comment of why we put the unlink then thanks !

File: addons/account/models/account_move_line.py (hunk 3 above)
Comment (chevalierv) [reply addons/account/models/account_move_line.py:3243]: The `len(self) == 1` is not really needed, we did that in Sales to avoid computing the price and sum the quantity.

Why did you remove the `self.product_id.ensure_one()' from your previous commit? In your use case, it's okay to sum the number of customizable desks and chairs? 

The catalog does not support editing products that are spread across multiple lines. We can't know on which line the customer wants to add or remove one item. The discount or the uom could also differ. In that case, in Sales, we show the price as if there was only 1 item in the order.

File: addons/account/models/account_move_line.py [hunk 6]
```diff
@@ -3232,6 +3232,51 @@ def action_automatic_entry(self, default_action=None):
...
+                **self[0].move_id._get_product_price_and_data(self[0].product_id),
+                'quantity': sum(
+                    self.mapped(
+                        lambda line: line.product_uom_id._compute_quantity(
+                            qty=line.quantity,
+                            to_unit=line.product_id.uom_id,
+                        )
+                    )
+                ),
+                'readOnly': self.move_id._is_readonly(),
```
Comment (chevalierv) [inline addons/account/models/account_move_line.py:3275]: In `_is_readonly`, you're checking that you have only one invoice. But here, we want to check that we only have one invoice line. As I stated in my previous comment, the catalog does not support editing products that are spread across multiple lines.
```suggestion
                'readOnly': len(self) > 1 or self.move_id._is_readonly(),
```

Comment (malb-odoo) [reply addons/account/models/account_move_line.py:3276]: We decided to put check on the length into the readonly function by checking the invoice_line_ids :+1:

File: addons/account/models/account_move.py [hunk 7]
```diff
@@ -2048,6 +2049,106 @@ def _validate_taxes_country(self):
...
+        res['product_catalog_currency_id'] = self.currency_id.id
+        res['product_catalog_digits'] = self.line_ids._fields['price_unit'].get_digits(self.env)
+        return res
+
+    def _get_product_catalog_domain(self):
+        if self.is_sale_document():
+            return expression.AND([super()._get_product_catalog_domain(), [('sale_ok', '=', True)]])
+        elif self.is_purchase_document():
+            return expression.AND([super()._get_product_catalog_domain(), [('purchase_ok', '=', True)]])
+        else: # In case of an entry
```
Comment (FlorianGilbert) [inline addons/account/models/account_move.py:2069]:
```suggestion
        else:  # In case of an entry
```
Cf: [PEP8](https://peps.python.org/pep-0008/#:~:text=Inline%20comments%20should%20be%20separated,a%20%23%20and%20a%20single%20space.)
:nerd_face:

File: addons/account/models/account_move.py [hunk 8]
```diff
@@ -2048,6 +2049,106 @@ def _validate_taxes_country(self):
...
+            return expression.AND([super()._get_product_catalog_domain(), []])
```
Comment (FlorianGilbert) [inline addons/account/models/account_move.py:2070]: Isn't it the same than this:
```suggestion
            return super()._get_product_catalog_domain()
```

File: addons/account/models/account_move.py [hunk 9]
```diff
@@ -2048,6 +2049,106 @@ def _validate_taxes_country(self):
                     raise ValidationError(_("This entry contains taxes that are not compatible with your fiscal position. Check the country set in fiscal position and in your tax configuration."))
                 raise ValidationError(_("This entry contains one or more taxes that are incompatible with your fiscal country. Check company fiscal country in the settings and tax country in taxes configuration."))
 
+    # -------------------------------------------------------------------------
+    # CATALOG
+    # -------------------------------------------------------------------------
+    def _get_action_add_from_catalog_extra_context(self):
+        res = super()._get_action_add_from_catalog_extra_context()
+        if not self.is_sale_document() and self.move_type != 'entry' and self.partner_id:
```
Comment (FlorianGilbert) [inline addons/account/models/account_move.py:2057]:
```suggestion
        if self.is_purchase_document() and self.partner_id:
```
What about this?

**Review by FlorianGilbert — CHANGES_REQUESTED** [review]
Almost good to me :)

**Review by FlorianGilbert — APPROVED** [review]
@robodoo r+

File: addons/account/models/account_move.py [hunk 10]
```diff
@@ -2048,6 +2049,112 @@ def _validate_taxes_country(self):
...
+        :param int quantity: The quantity selected in the catalog
+        :return: The unit price of the product, based on the pricelist of the
+                 sale order and the quantity selected.
+        :rtype: float
+        """
+        move_line = self.line_ids.filtered(lambda line: line.product_id.id == product_id)
+        if move_line:
+            if quantity != 0:
+                move_line.quantity = quantity
+            elif self.state in {'draft', 'sent'}:
```
Comment (william-andre) [inline addons/account/models/account_move.py:2134]: Yop, I'm checking my linter and see that this `sent` state doesn't exist.
Should I just remove of check `is_move_sent` instead? :thinking:

Comment (malb-odoo) [reply addons/account/models/account_move.py:2134]: I think it's a bad copy paste form sale_order :eyes: Since we don't have that on move maybe we can just put self.state == draft
